File: ios/SampleUI/Sources/SampleUI/State/UseSmileIDSampleTokenSession.swift
import Foundation

/// A linked session, held as an absolute deadline because a counter restarts wrong after process death.
public struct UseSmileIDSampleTokenSession: Equatable, Sendable, CustomStringConvertible {
  /// A display handle — the token's `jti`, else a digest of it. Never a prefix of the credential.
  public let id: String
  public let token: String
  public let issuedAt: Date
  public let expiresAt: Date
  public let bindings: UseSmileIDSampleTokenBindings
  /// The partner from the token's own claim, never a locally configured id, which is how a signed request gets a 401. Never logged.
  public let partnerId: String?
  /// From the token's own `api_url` claim. Non-nil by construction: the decoder refuses a token it cannot place.
  public let environment: UseSmileIDSampleEnvironment

  public init(
    id: String,
    token: String,
    issuedAt: Date,
    expiresAt: Date,
    bindings: UseSmileIDSampleTokenBindings,
    partnerId: String? = nil,
    environment: UseSmileIDSampleEnvironment
  ) {
    self.id = id
    self.token = token
    self.issuedAt = issuedAt
    self.expiresAt = expiresAt
    self.bindings = bindings
    self.partnerId = partnerId
    self.environment = environment
  }

  public func remaining(at now: Date) -> TimeInterval {
    max(0, expiresAt.timeIntervalSince(now))
  }

  public func hasExpired(at now: Date) -> Bool {
    now >= expiresAt
  }

  /// 1 fresh down to 0 expired, over the token's own span; the zero-span guard is for the public initialiser, since NaN survives a clamp.
  public func progress(at now: Date) -> Double {
    let span = max(expiresAt.timeIntervalSince(issuedAt), 0.001)
    return min(max(remaining(at: now) / span, 0), 1)
  }

  /// Redacted: a synthesised description is how a bearer credential reaches a log or a crash report.
  public var description: String {
    "UseSmileIDSampleTokenSession(id=\(id), environment=\(environment.label), expiresAt=\(expiresAt.timeIntervalSince1970))"
  }
}

/// `m:ss`, growing an hours part when the span needs one (7:59:12, not 479:12); floors, so it never shows time that has gone.
public func useSmileIDSampleCountdown(_ remaining: TimeInterval) -> String {
  let total = Int(max(0, remaining).rounded(.down))
  let hours = total / 3600
  let minutes = (total % 3600) / 60
  let seconds = total % 60
  if hours > 0 {
    return "\(hours):\(padded(minutes)):\(padded(seconds))"
  }
  return "\(minutes):\(padded(seconds))"
}

private func padded(_ value: Int) -> String {
  value < 10 ? "0\(value)" : "\(value)"
}

File: expo/sample-ui/src/state/use-smile-id-sample-token-session.ts
import type { UseSmileIDSampleEnvironment } from '../model/use-smile-id-sample-result';
import type { UseSmileIDSampleTokenBindings } from './use-smile-id-sample-token-decoder';

/// A linked session, held as an absolute deadline because a counter restarts wrong after process death. Only the decoder builds one.
export type UseSmileIDSampleTokenSession = {
  /// A display handle — the token's `jti`, else a digest of it. Never a prefix of the credential.
  readonly id: string;
  readonly token: string;
  readonly issuedAtMillis: number;
  readonly expiresAtMillis: number;
  readonly bindings: UseSmileIDSampleTokenBindings;
  /// From the token's own `partner_id`, which wins over the profile. Never logged.
  readonly partnerId: string | null;
  /// From the token's own `api_url`; the decoder refuses a token it cannot place.
  readonly environment: UseSmileIDSampleEnvironment;
};

/// A session that ran out, remembered without its credential.
export type UseSmileIDSampleEndedSession = {
  readonly id: string;
  readonly endedAtMillis: number;
};

/// Milliseconds left, never negative.
export const smileIDSampleSessionRemaining = (session: UseSmileIDSampleTokenSession, nowMillis: number): number =>
  Math.max(session.expiresAtMillis - nowMillis, 0);

/// Expiry is the deadline itself.
export const smileIDSampleSessionHasExpired = (session: UseSmileIDSampleTokenSession, nowMillis: number): boolean =>
  nowMillis >= session.expiresAtMillis;

/// 1 fresh to 0 expired over the token's own span; a zero span reads as spent rather than NaN.
export const smileIDSampleSessionProgress = (session: UseSmileIDSampleTokenSession, nowMillis: number): number => {
  const span = Math.max(session.expiresAtMillis - session.issuedAtMillis, 1);
  return Math.min(Math.max(smileIDSampleSessionRemaining(session, nowMillis) / span, 0), 1);
};

const pad = (value: number) => String(value).padStart(2, '0');

/// `m:ss`, growing an hours part when the span needs one — an 8h token reads 7:59:12, not 479:12.
export const smileIDSampleCountdown = (remainingMillis: number): string => {
  const total = Math.floor(Math.max(remainingMillis, 0) / 1000);
  const hours = Math.floor(total / 3600);
  const minutes = Math.floor((total % 3600) / 60);
  const seconds = total % 60;
  return hours > 0 ? `${hours}:${pad(minutes)}:${pad(seconds)}` : `${minutes}:${pad(seconds)}`;
};
